File: ports/go/src/core/tools/bcsvariant.go
package tools

import (
	"fmt"
	"reflect"
)

// BcsVariant is a type-erased container, conceptually mapping to `interface{}` in Go
type BcsVariant struct {
	value interface{}
}

func NewBcsVariant(val interface{}) *BcsVariant {
	return &BcsVariant{value: val}
}

// Value retrieves the underlying value
func (v *BcsVariant) Value() interface{} {
	return v.value
}

// SetValue assigns a new type-erased value
func (v *BcsVariant) SetValue(val interface{}) {
	v.value = val
}

// IsNil checks if the variant has no underlying type/value
func (v *BcsVariant) IsNil() bool {
	return v.value == nil
}

// GetType retrieves the string representation of the underlying type
func (v *BcsVariant) GetType() string {
	if v.IsNil() {
		return "Nil"
	}
	return reflect.TypeOf(v.value).String()
}

func (v *BcsVariant) ToString() string {
	return fmt.Sprintf("%v", v.value)
}

func (v *BcsVariant) ToInt() (int, bool) {
	if val, ok := v.value.(int); ok {
		return val, true
	}
	return 0, false
}

func (v *BcsVariant) ToFloat() (float64, bool) {
	if val, ok := v.value.(float64); ok {
		return val, true
	}
	return 0.0, false
}

func (v *BcsVariant) ToBool() (bool, bool) {
	if val, ok := v.value.(bool); ok {
		return val, true
	}
	return false, false
}

File: ports/rust/src/core/tools/bcsvariant.rs
// BcsVariant provides a type-erased enum for dynamic properties in Rust
#[derive(Clone, Debug)]
pub enum BcsVariant {
    Int(i32),
    Float(f64),
    String(String),
    Bool(bool),
    Nil,
}

impl BcsVariant {
    pub fn new_int(val: i32) -> Self {
        BcsVariant::Int(val)
    }

    pub fn new_float(val: f64) -> Self {
        BcsVariant::Float(val)
    }

    pub fn new_string(val: &str) -> Self {
        BcsVariant::String(val.to_string())
    }

    pub fn new_bool(val: bool) -> Self {
        BcsVariant::Bool(val)
    }

    pub fn is_nil(&self) -> bool {
        matches!(self, BcsVariant::Nil)
    }

    pub fn get_type(&self) -> String {
        match self {
            BcsVariant::Int(_) => "Int".to_string(),
            BcsVariant::Float(_) => "Float".to_string(),
            BcsVariant::String(_) => "String".to_string(),
            BcsVariant::Bool(_) => "Bool".to_string(),
            BcsVariant::Nil => "Nil".to_string(),
        }
    }

    pub fn to_string(&self) -> String {
        match self {
            BcsVariant::Int(v) => v.to_string(),
            BcsVariant::Float(v) => v.to_string(),
            BcsVariant::String(v) => v.clone(),
            BcsVariant::Bool(v) => v.to_string(),
            BcsVariant::Nil => "Nil".to_string(),
        }
    }

    pub fn to_int(&self) -> Option<i32> {
        if let BcsVariant::Int(v) = self {
            Some(*v)
        } else {
            None
        }
    }

    pub fn to_float(&self) -> Option<f64> {
        if let BcsVariant::Float(v) = self {
            Some(*v)
        } else {
            None
        }
    }

    pub fn to_bool(&self) -> Option<bool> {
        if let BcsVariant::Bool(v) = self {
            Some(*v)
        } else {
            None
        }
    }
}
